### PR TITLE
editoast: Setup sentry

### LINKED ISSUE #3178: Setup sentry for editoast
### Description and goal

Editoast errors, should be reported on sentry. Here the documentation to do so with actix-web: https://docs.sentry.io/platforms/rust/guides/actix-web/

### Acceptance criteria

- Errors are reported on sentry on dev env.

### PR DESCRIPTION
**Pull request by bloussou**
Setup sentry for errors for editoast. 

Close https://github.com/DGEXSolutions/osrd/issues/3178

### CODE REVIEW

File: editoast/src/main.rs [hunk 1]
```diff
@@ -86,6 +86,18 @@ async fn runserver(
     // Setup shared states
     let infra_caches = Data::new(CHashMap::<i64, InfraCache>::default());
 
+    // Setup sentry
+    if args.sentry_dsn.is_some() && args.sentry_env.is_some() {
```
Comment (multun) [inline editoast/src/main.rs]: can you pattern-match instead, and display a warning if either is set but not both?
this way, if someone sets sentry_dsn but not sentry_env, if would immediately be obvious that something is off

File: editoast/src/main.rs [hunk 2]
```diff
@@ -86,6 +86,18 @@ async fn runserver(
...
+        let _guard = sentry::init((
```
Comment (multun) [inline editoast/src/main.rs]: I believe the guard should only be dropped when the scope of main dies. Please wrap the guard in an Option and let it live as long as the program. You could also move this to a subroutine

**Review by multun — CHANGES_REQUESTED** [review]

File: editoast/src/main.rs (hunk 2 above)
Comment (bloussou) [reply editoast/src/main.rs]: looks like it is working this way with some unsafe code in the crate :confused:

**Review by flomonster — DISMISSED** [review]

File: editoast/src/main.rs [hunk 3]
```diff
@@ -86,6 +87,26 @@ async fn runserver(
     // Setup shared states
     let infra_caches = Data::new(CHashMap::<i64, InfraCache>::default());
 
+    // Setup sentry
+    let _guard = match (
+        args.sentry_dsn,
+        args.sentry_env,
+        env::var("OSRD_GIT_DESCRIBE").ok(),
```
Comment (multun) [inline editoast/src/main.rs]: OSRD_GIT_DESCRIBE is optional

Comment (Khoyo) [reply editoast/src/main.rs]: I think it would be reasonable to require it to send data to sentry.

File: editoast/src/main.rs [hunk 4]
```diff
@@ -86,6 +87,26 @@ async fn runserver(
...
+        (Some(sentry_dsn), Some(sentry_env), Some(release)) => Some(sentry::init((
+            sentry_dsn,
+            sentry::ClientOptions {
+                release: Some(release.into()),
+                environment: Some(sentry_env.into()),
+                ..Default::default()
+            },
+        ))),
+        _ => {
+            println!("Missing SENTRY_DSN or SENTRY_ENV or OSRD_GIT_DESCRIBE env variables to sent event to sentry.");
```
Comment (Khoyo) [inline editoast/src/main.rs]:
```suggestion
            println!("Missing SENTRY_DSN, SENTRY_ENV or OSRD_GIT_DESCRIBE env variable to send events to sentry.");
```

File: editoast/src/main.rs (hunk 2 above)
Comment (multun) [reply editoast/src/main.rs]: > The [sentry::init](https://docs.rs/sentry/latest/sentry/fn.init.html) function returns a guard that when dropped will flush Events that were not yet sent to the sentry service. It has a two second deadline for this so shutdown of applications might slightly delay as a result of this. Keep the guard around or sending events will not work.

https://docs.rs/sentry/latest/sentry/

File: editoast/src/main.rs [hunk 5]
```diff
@@ -60,6 +62,26 @@ async fn run() -> Result<(), Box<dyn Error + Send + Sync>> {
...
 
+fn init_sentry(args: &RunserverArgs) -> Option<ClientInitGuard> {
+    match (args.sentry_dsn.clone(), args.sentry_env.clone()) {
+        (Some(sentry_dsn), Some(sentry_env)) => Some(sentry::init((
+            sentry_dsn,
+            sentry::ClientOptions {
+                release: match env::var("OSRD_GIT_DESCRIBE").ok() {
+                    Some(release) => Some(release.into()),
+                    None => sentry::release_name!(),
+                },
```
Comment (multun) [inline editoast/src/main.rs:73]:
```suggestion
                release: env::var("OSRD_GIT_DESCRIBE").unwrap_or_else(|| sentry::release_name!()),
```

File: editoast/src/main.rs [hunk 6]
```diff
@@ -60,6 +62,26 @@ async fn run() -> Result<(), Box<dyn Error + Send + Sync>> {
...
+                environment: Some(sentry_env.into()),
+                ..Default::default()
+            },
+        ))),
+        _ => {
+            println!("SENTRY_DSN and SENTRY_ENV must be set to send events to Sentry.");
+            None
```
Comment (multun) [inline editoast/src/main.rs]: If both SENTRY_DSN and SENTRY_ENV aren't set, I think it's safe to say the user didn't intend to use sentry, and thus that no warning should be displayed

File: editoast/src/main.rs [hunk 7]
```diff
@@ -94,6 +119,7 @@ async fn runserver(
             .allow_any_header();
 
         App::new()
+            .wrap(sentry_actix::Sentry::new())
```
Comment (multun) [inline editoast/src/main.rs]: shouldn't this only be added if sentry was initialized?

**Review by multun — CHANGES_REQUESTED** [review]

File: editoast/src/main.rs (hunk 5 above)
Comment (bloussou) [reply editoast/src/main.rs:73]: Not sure I can have a simple one liner since it is not a `String` but a `std::borrow::Cow` that is expected

Comment (multun) [reply editoast/src/main.rs:73]: okaaay, gotcha

**Review by multun — APPROVED** [review]

**Review by Tguisnet — APPROVED** [review]

**Pull request merged**